### DISCUSSION TITLE
How to create a nlp object with a dict instead of a string?

### DISCUSSION THREAD

**Discussion by lingvisa** [explosion/spaCy#10129] [Help: Other Questions] [answered]
The typical way of creating a nlp and doc object is:

```
nlp = spacy.load("en_core_web_sm")
doc = nlp("Apple is looking at buying U.K. startup for $1 billion")
```

Is it possible that my input is not a str, but a dict like:

`input = {'text': 'text', 'summary': 'summary', 'id': '001', 'type': 'News'}`

Is it possible that a doc can created like:

`doc = nlp(input)`

The reason is that, my  custom component needs to access other fields of the input, beyond the usual 'text' field.  Please advise ideas if this is possible.

**Comment by polm** [comment, marked as answer]
You can't exactly pass a dict as input, but in 3.2 we added a feature for exactly this kind of use case such that you can [use a Doc as input to a pipeline](https://spacy.io/usage/v3-2#doc-input) (pipeline = nlp object). So what you can do is have a minimal pipeline that does maybe just tokenization, then add your data to the Doc as underscore attributes, and then pass that to another pipeline with your custom components.

**Comment by lingvisa** [comment]
Good to know the new features. What about the component_cfg = {} argument? Can I access this argument from a nlp object?

**Reply by polm** [reply]
Yes, the contents of `component_cfg` are passed to the respective components as keyword arguments.

**Reply by lingvisa** [reply]
However, how can I get the component_cfg from a pipeline (nlp) or a doc within my custom component? Something like:
nlp.component_cfg or doc.component_cfg? Does component_cfg and 'use Doc as input to a pipeline' serve similar purposes?

**Reply by lingvisa** [reply]
> Yes, the contents of `component_cfg` are passed to the respective components as keyword arguments.

Please tell how to access the cocmponent_cfg's contents in a component? Thanks.

**Reply by polm** [reply]
Here's an example of using `component_cfg`. 

```
import spacy
from spacy.tokens import Doc
from spacy.language import Language

nlp = spacy.blank("en")
Doc.set_extension("bloop", default=None)

@Language.component("bloop")
def bloop(doc, msg=None):
    if msg is not None:
        doc._.bloop = msg
        print(msg)
    return doc

nlp.add_pipe("bloop")

# prints nothing
doc = nlp("I like peanut butter")
# is None
print(doc._.bloop)
# prints "hello there"
doc = nlp("I like sushi", component_cfg={"bloop":{"msg":"hello there"}})
# is "hello there"
print(doc._.bloop)
```

The top level keys in `component_cfg` correspond to component names, and the contents are passed as keyword arguments to the component.

**Reply by lingvisa** [reply]
If bloop is a component class, then it should be like this :
```
 @Language.factory("bloop")
def create_bloop(nlp:Language, name:str):
    return Bloop(nlp)

class Bloop(object):
   def __init__(nlp:Language):
      ...

  def __call__(self, doc, user_args):
     print(user_args)
```
Then I can do:
`doc = nlp("I like sushi", component_cfg={"bloop":{"user_args":{"test1":"hello", "test2":{"test3":"hello2"}}}})`

This looks like exactly what I need.  I tested and it worked! Thanks. I don't have to upgrade to spacy 3.2 for the time being.

**Reply by polm** [reply]
It might be helpful to realize the that the `component_cfg` is just the same as what you would put in a config file - check [the docs](https://spacy.io/usage/processing-pipelines#custom-components-factories) on how the config is loaded. The values in the config are basically passed in the same way (though at the point the component is created, not on individual calls).

**Reply by lingvisa** [reply]
That's true. Thanks.
